runner/src/runner.rs

```diff
@@ -100,23 +100,207 @@ impl Runner {
+        )
+        .map_err(|_| Error::ProcessTerminationFailure)?;
+
+        match child_proc.wait() {
+            Ok(exit_status) => log::info!(
+                "Terminated vault process (pid: {}) with exit status {}",
+                child_proc.id(),
+                exit_status
+            ),
+            Err(error) => log::warn!("Vault process termination error: {}", error),
+        };
+        let pid = child_proc.id();
+        runner.set_child_proc(None);
+        Ok(pid)
+    }
+
+    async fn try_get_release<T: RunnerExt + StorageReader>(
+        runner: &T,
+        pending: bool,
+    ) -> Result<Option<ClientRelease>, Error> {
+        let storage_item = if pending {
+            PENDING_RELEASE_STORAGE_ITEM
+        } else {
+            CURRENT_RELEASE_STORAGE_ITEM
+        };
+
+        let storage_key = runner.compute_storage_key(PARACHAIN_MODULE.to_string(), storage_item.to_string());
+
+        retry_with_log_async(
+            || {
+                let key = &storage_key;
+                runner
+                    .read_chain_storage::<ClientRelease>(runner.parachain_rpc(), Some(key.to_string()))
+                    .into_future()
+                    .boxed()
+            },
+            "Error fetching executable".to_string(),
+        )
+        .await
+    }
+
+    async fn read_chain_storage<T: 'static + Decode + Debug, V: RunnerExt + StorageReader>(
+        runner: &V,
+        parachain_rpc: &WebsocketClient,
+        maybe_storage_key: Option<String>,
+    ) -> Result<Option<T>, Error> {
+        let enc_res = runner
+            .query_storage(parachain_rpc, maybe_storage_key, "state_getStorage".to_string())
+            .await;
+        enc_res
+            .map(|r| T::decode(&mut r.as_bytes_ref()))
+            .transpose()
+            .map_err(Into::into)
+    }
+
+    pub async fn run(mut runner: Box<dyn RunnerExt + Send>, mut shutdown_signals: SignalsInfo) -> Result<(), Error> {
+        tokio::select! {
+            _ = shutdown_signals.next() => {
+                log::info!("Terminating vault...");
+                runner.terminate_proc_and_wait()?;
+            },
+            _ = runner.as_mut().auto_update() => {
+                log::error!("Auto-updater unexpectedly terminated.");
+                return Err(Error::AutoUpdaterTerminated);
+            }
+        };
+        Ok(())
+    }
+
+    async fn auto_update(runner: &mut impl RunnerExt) -> Result<(), Error> {
+        // Create all directories for the `download_path` if they don't already exist.
+        fs::create_dir_all(&runner.download_path())?;
+        let release = runner.try_get_release(false).await?.expect("No current release");
+        // WARNING: This will overwrite any pre-existing binary with the same name
+        // TODO: Check if a release with the same version is already at the `download_path`
+        runner.download_binary(release).await?;
+
+        runner.run_binary()?;
+
+        loop {
```

Comment:
Shouldn't we keep trying if we get an error in this loop? Or do we leave it do e.g. systemd to automatically restart upon failure?

Reply:
My assumption was that all transient errors that may occur are already wrapped in retries, so any failure in this loop is unfixable. We should still run it as a systemd service in case any edge case was missed.

Reply:
They are wrapped in retries, but these only last for a minute, right? A network outage could easily last longer than that. It's ok if we decide to rely on systemd but it needs to be a conscious decision. What happens when we quit the runner if the vault still runs? Would that lead to problem on startup, starting a second vault in parallel?

Also, what about the reverse? What happens when the vault client terminates? I don't think we catch that, right? So if e.g. the bitcoin node goes down, causing the vault to quit (eventually), the runner would keep on running uselessly? If we use systemd to start the runner, we need to be able to assume that if the runner runs, then the vault runs.

Reply:
> A network outage could easily last longer than that. 
Ultimately this could be configured by the user. What does the Bitcoin node do on a network timeout?

> What happens when we quit the runner if the vault still runs?
This was covered in PR https://github.com/interlay/interbtc-clients/pull/330

> What happens when the vault client terminates?
This is indeed not handled, a check for the status of the child process should be included in the loop in case it needs to be restarted. Currently the runner just keeps running uselessly until a new release is found.

Reply:
> This was covered in PR https://github.com/interlay/interbtc-clients/pull/330
I don't think it was? I think that was only catching explicit kill signals, e.g. the user pressing ctrl+c. But if on of the functions below returns an error, the runner will also exit without receiving any signal, right?